### PR TITLE
[SPARK-48712][SQL][FOLLOWUP] Check whether input is valid utf-8 string or not before entering fast path

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

Check whether input is valid utf-8 string or not before entering fast path

### Why are the changes needed?

Avoid behavior change on a corner case where users provide invalid UTF-8 strings for UTF-8 encoding

### Does this PR introduce _any_ user-facing change?
no, this is a followup to avoid potential breaking change

### How was this patch tested?
existing tests

### Was this patch authored or co-authored using generative AI tooling?
no

### CODE REVIEW

**Comment by yaooqinn** [comment]
cc @cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala [hunk 1]
```diff
@@ -3039,7 +3039,7 @@ object Encode {
       legacyCharsets: Boolean,
       legacyErrorAction: Boolean): Array[Byte] = {
     val toCharset = charset.toString
-    if (input.numBytes == 0 || "UTF-8".equalsIgnoreCase(toCharset)) {
+    if (input.numBytes == 0 || ("UTF-8".equalsIgnoreCase(toCharset) && input.isValid())) {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:3042]: can we also add a test for invalid utf8 input?

Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:3042]: Ok, sure

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:3042]: BTW let's also test empty bytes. Did we fail before this optimization?

Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:3042]: I can't create a 'failed case before or after this PR. Invalid inputs do not always generate errors, but a re-mapped byte array. It's tricky for me to find a specific case to test `CharacterCodingException`

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/StringExpressionsSuite.scala [hunk 2]
```diff
@@ -2076,4 +2077,14 @@ class StringExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper {
...
   }
+
+  test("SPARK-48712: Check whether input is valid utf-8 string or not before entering fast path") {
+    val str = UTF8String.fromBytes(Array[Byte](-1, -2, -3, -4))
+    assert(!str.isValid, "please use a string that is not valid UTF-8 for testing")
+    val expected = Array[Byte](-17, -65, -67, -17, -65, -67, -17, -65, -67, -17, -65, -67)
+    val bytes = Encode.encode(str, UTF8String.fromString("UTF-8"), false, false)
+    assert(bytes === expected)
+    checkEvaluation(Encode(Literal(str), Literal("UTF-8")), expected)
+    checkEvaluation(Encode(Literal(UTF8String.EMPTY_UTF8), Literal("UTF-8")), Array.emptyByteArray)
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/StringExpressionsSuite.scala:2088]: `EMPTY_UTF8` is different from an empty byte array, can we test `UTF8String.fromBytes(Array[Byte]())`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala [hunk 3]
```diff
@@ -3039,7 +3039,7 @@ object Encode {
       legacyCharsets: Boolean,
       legacyErrorAction: Boolean): Array[Byte] = {
     val toCharset = charset.toString
-    if (input.numBytes == 0 || "UTF-8".equalsIgnoreCase(toCharset)) {
+    if (input.numBytes == 0 || ("UTF-8".equalsIgnoreCase(toCharset) && input.isValid)) {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:3042]: another missed case is when the charset name is invalid. Previously we always fail even though the input bytes is empty.

Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:3042]: make sense

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala [hunk 4]
```diff
@@ -3039,16 +3039,20 @@ object Encode {
...
-      return input.getBytes
-    }
-    val encoder = CharsetProvider.newEncoder(toCharset, legacyCharsets, legacyErrorAction)
-    try {
-      val bb = encoder.encode(CharBuffer.wrap(input.toString))
-      JavaUtils.bufferToArray(bb)
-    } catch {
-      case _: CharacterCodingException =>
-        throw QueryExecutionErrors.malformedCharacterCoding("encode", toCharset)
+    if ("UTF-8".equalsIgnoreCase(toCharset) && input.isValid) {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:3042]: Personally I feel code with shortcuts is easier to read. How about
```
if ("UTF-8".equalsIgnoreCase(toCharset) && input.isValid) return input.getBytes
val encoder = CharsetProvider.newEncoder(toCharset, legacyCharsets, legacyErrorAction)
if (input.numBytes == 0) return input.getBytes
...
```

**Review by cloud-fan — APPROVED** [review]

**Comment by yaooqinn** [comment]
Merged into master. Thanks, @cloud-fan, for raising the issue and providing a thorough review.

**Pull request closed without merging**